### PR TITLE
Enable providers to use other providers from application libctx

### PR DESCRIPTION
**Pull request by mattcaswell**
This is an attempt to fix the issues discussed in #14925.

A provider may want to access algorithms from other providers. Therefore some providers may just want to use the same libctx as used by the application - and therefore reuse all the providers that have already been loaded there.

However a provider may be linked against a different libcrypto to the application (e.g. if one is using static linking and another is using dynamic linking). Therefore it is not possible for the provider to use exactly the *same* libctx as the application.

To get around this we introduce the concept of a "child" libctx. The provider creates a child libctx which is linked to the application's libctx. All the provider's loaded into the application's libctx are automatically mirrored in the child libctx. This works by copying the provctx and OSSL_DISPATCH table used by the "parent" providers in the "parent" libctx into the child providers in the child libctx.

As far as the provider is concerned it appears as if it is using the "same" libctx. Providers are not reloaded and can be seemly reused across the parent and child libctxs.

### CODE REVIEW

**Comment by mattcaswell** [comment]
This is now essentially working except:

1) The test I added is currently failing due to #15032 

2) I have not attempted to synchronise the global properties between the parent and child libctx. I think this should be relatively straight forward.

3) Needs documentation updates

Before getting this PR into a reviewable state, #14925 needs review by OTC in light of this PR, to check we are going in the right direction.

**Comment by mattcaswell** [comment]
Lots of updates to this to properly handle changes in the loaded providers, add tests and missing documentation. This is now essentially finished except it depends on and includes the commits from #15042. I will keep this as WIP until then but is otherwise reviewable.

This does *not* include the ability to track changes in the application's global properties. I will handle that in a separate follow-on PR. This PR is already big enough.

File: crypto/provider_core.c
Comment (paulidale) [inline crypto/provider_core.c:105]: Might this be better conditioned out under FIPS_MODULE.

File: crypto/provider_core.c
Comment (paulidale) [inline crypto/provider_core.c:766]: Release the `store->lock` before returning.

File: crypto/provider_core.c
Comment (paulidale) [inline crypto/provider_core.c:819]: Release `store->lock` before returning.

File: crypto/provider_core.c
Comment (paulidale) [inline crypto/provider_core.c:862]: You might need a rebase.  This needs to be done under a read lock to avoid different bit fields fighting over the memory location.

File: crypto/provider_core.c
Comment (paulidale) [inline crypto/provider_core.c:1207]: Definitely need a rebase, this was changed to a read lock.

File: crypto/provider_core.c
Comment (paulidale) [inline crypto/provider_core.c:1393]: What about using `sk_OSSL_PROVIDER_CHILD_CB_find` ?

File: test/p_test.c
Comment (paulidale) [inline test/p_test.c:127]: This will need conditioning on `OPENSSL_NO_MD4`

File: test/p_test.c
Comment (paulidale) [inline test/p_test.c:137]: Is diagnostic output possible here to distinguish the failure cases?

File: include/internal/cryptlib.h
Comment (paulidale) [inline include/internal/cryptlib.h:176]: I generally prefer float for priorities (they allow insertion between existing values) but since this is internal and we control everything, I don't see a need.

**Review by paulidale — CHANGES_REQUESTED** [review]
This looks quite good.

File: crypto/provider_core.c
Comment (mattcaswell) [reply crypto/provider_core.c:1464]: `sk_OSSL_PROVIDER_CHILD_CB_find` helps you find a specific entry in the stack. Here we need to loop over all of them.

File: test/p_test.c
Comment (mattcaswell) [reply test/p_test.c:138]: I'm not sure how to emit diagnostic output from inside this test provider?

**Comment by mattcaswell** [comment]
All feedback addressed or responded to above. I've rebased on the latest version of #15042 and will rebase again once that is merged.

I also fixed a nasty bug relating to built-in providers.

File: test/p_test.c
Comment (paulidale) [reply test/p_test.c:138]: What about raising an error?

Comment (mattcaswell) [reply test/p_test.c:138]: I've raised different error codes for different types of failure cases.

**Comment by mattcaswell** [comment]
The final outstanding comment has been resolved and I've rebased to set the latest fips checksums to avoid conflicts with master.

File: crypto/ex_data.c
Comment (paulidale) [inline crypto/ex_data.c:354]: What about `return bp->excb->priority - ap->excb->priority;` instead of this and the subsequent two lines?

File: crypto/provider_child.c
Comment (paulidale) [inline crypto/provider_child.c:82]: This will come back to bite us ....

File: crypto/provider_child.c
Comment (paulidale) [inline crypto/provider_child.c:127]: Spurious blank line.

**Review by paulidale — APPROVED** [review]

**Pull request closed without merging**
